flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -5366,27 +5448,123 @@ class ArrayExprLowering {
     TODO(getLoc(), "coarray ref");
   }
 
+  CC genarr(const Fortran::evaluate::StaticDataObject::Pointer &) {
+    fir::emitFatalError(getLoc(), "substring of static array object");
+  }
+
   /// 9.4.1 Substrings
   CC genarr(const Fortran::evaluate::Substring &x) {
     auto loc = getLoc();
-    auto pf = std::visit(
-        Fortran::common::visitors{
-            [&](const Fortran::evaluate::DataRef &p) { return genarr(p); },
-            [&](const Fortran::evaluate::StaticDataObject::Pointer &) -> CC {
-              fir::emitFatalError(loc, "substring of static array object");
-            }},
-        x.parent());
-    // lower and upper *must* be scalars
-    llvm::SmallVector<mlir::Value> bounds = {fir::getBase(asScalar(x.lower()))};
-    if (auto upper = x.upper())
-      bounds.push_back(fir::getBase(asScalar(*upper)));
-    return [=](IterSpace iters) -> ExtValue {
+    if (isProjectedCopyInCopyOut()) {
+      ccStoreToDest = [=, &x](IterSpace iters) {
+        auto loc = getLoc();
+        auto innerArg = iters.innerArgument();
+        auto exv = iters.elementExv();
+        auto arrTy = innerArg.getType();
+        auto eleTy = fir::applyPathToType(arrTy, iters.iterVec());
+        auto resRefTy = builder.getRefType(eleTy);
+        // Get a reference to the array element to be amended.
+        auto arrayOp = builder.create<fir::ArrayAccessOp>(
+            loc, resRefTy, innerArg, iters.iterVec(), destination.typeparams());
+        auto charTy = eleTy.dyn_cast<fir::CharacterType>();
+        assert(charTy && "must be character");
+        llvm::SmallVector<mlir::Value> bounds = {
+            fir::getBase(asScalar(x.lower()))};
+        if (auto upper = x.upper())
+          bounds.push_back(fir::getBase(asScalar(*upper)));
+        fir::factory::CharacterExprHelper helper{builder, loc};
+        auto len = fir::getLen(exv);
+        auto charBox =
+            helper.createSubstring(fir::CharBoxValue{arrayOp, len}, bounds);
+        // Generate an optimized CHARACTER assignment.
+        fir::factory::genCharacterCopy(charBox.getAddr(), charBox.getLen(),
+                                       fir::getBase(exv), fir::getLen(exv),
+                                       builder, loc);
+        // Mark this array element as amended.
+        auto ty = innerArg.getType();
+        auto amend = builder.create<fir::ArrayAmendOp>(loc, ty, innerArg,
+                                                       charBox.getAddr());
+        return abstractArrayExtValue(amend, len);
+      };
+    }
+    auto pf = std::visit([&](const auto &v) { return genarr(v); }, x.parent());
+    if (!isBoxValue())
+      return pf;
```

Comment:
I think this is now bypassing the substring for the CustomCopyInCopyOut.
I find it a bit weird that we are applying the constituents semantics at that stage. Would it be possible to instead have the substring in cmptData like for trips and components ?